### ISSUE TITLE
BleachBit v4.4.0 display unreadble text after installed on win7(64bit), Simplified Chinese os

### ISSUE THREAD

**Issue by huastone2018** [bleachbit/bleachbit#1207]
**Describe the bug**
The newest BleachBit v4.4.0 display unreadble text when I installed it on win7(64bit), Simplified Chinese operating system. Please see the screenshots below. I like this tools and expect this could be resolved, thank you.

**To Reproduce**

1. install the Bleachbit new version v4.40
2. It prompt old version BleachBit already exist, and follow the instructions to unstall the old one(v4.0.0)
3. continue to install BleachBit v4.4.0
4. run the BleachBit by click from the start menu on win7 64bit, Simplified Chinese operating system.
5. It display unreadble text srceen after the window opened.

The error happens (sometimes/every time).

**Expected behavior**
Tt should display nornal Simplified Chinese screen just like the install process.

**Screenshots**
![bleachbit bug1](https://user-images.githubusercontent.com/39025953/124053594-21414b00-da53-11eb-89d9-1edc1eb1bdef.png)

**Desktop (please complete the following information):**
 - OS: win7 bit64
 - BleachBit version 4.4.0

**Additional context**
System info displayed by BleachBit is list below by texts:
BleachBit version 4.4.0
Build number 1929
GTK version 3.18.9
GTK theme = Adwaita
GTK icon theme = Adwaita
GTK prefer dark theme = True
SQLite version 3.35.0
local_cleaners_dir = None
locale_dir = C:\Program Files (x86)\BleachBit\share\locale\
options_dir = C:\Users\rs\AppData\Roaming\BleachBit
personal_cleaners_dir = C:\Users\rs\AppData\Roaming\BleachBit\cleaners
system_cleaners_dir = C:\Program Files (x86)\BleachBit\share\cleaners\
locale.getdefaultlocale = ('zh_CN', 'cp936')
os.getenv('APPDATA') = C:\Users\rs\AppData\Roaming
os.getenv('cd') = C:\Program Files (x86)\BleachBit
os.getenv('LocalAppData') = C:\Users\rs\AppData\Local
os.getenv('LocalAppDataLow') = C:\Users\rs\AppData\LocalLow
os.getenv('Music') = C:\Users\rs\Music
os.getenv('USERPROFILE') = C:\Users\rs
os.getenv('ProgramFiles') = C:\Program Files (x86)
os.getenv('ProgramW6432') = C:\Program Files
os.getenv('TMP') = C:\Users\rs\AppData\Local\Temp
os.path.expanduser('~') = C:\Users\rs
platform.win32_ver[1]() = 6.1.7601
platform.platform = Windows-7-6.1.7601-SP1
platform.version = 6.1.7601
sys.argv = ['C:\\Program Files (x86)\\BleachBit\\bleachbit.exe']
sys.executable = C:\Program Files (x86)\BleachBit\bleachbit.exe
sys.version = 3.4.4 (v3.4.4:737efcadf5a6, Dec 20 2015, 19:28:18) [MSC v.1600 32 bit (Intel)]
win32com.shell.shell.IsUserAnAdmin() = False
__file__ = C:\Program Files (x86)\BleachBit\library.zip\bleachbit\SystemInformation.pyo

**Comment by az0** [comment]
@huastone2018 
Thank you for a bug report with excellent detail. Would you please also check one more thing? In the folder `c:\program files (x86)\bleachbit\share\fonts` are there the two fonts segoeui and tahoma?

**Comment by az0** [comment]
For comparison, here is a screenshot of my fonts folder
![share_fonts](https://user-images.githubusercontent.com/933680/124062339-cc1f1e00-d9ed-11eb-9021-87dcc5660528.png)

**Comment by huastone2018** [comment]
Thanks for your attention. Segoeui and tahoma exist in the folder you specified when bleachBit v4.4.0 installed. But them do not exist in the folder when BleachBit v4.0.0 installed. It display unreadable text on v4.4.0 and normal on v4.0.0. Please see the screenshot below:
![fonts screenshot when beachBit 4 4 0](https://user-images.githubusercontent.com/39025953/124062346-2c03dc00-da63-11eb-8ee2-20d28e4bad4a.png)
![fonts screenshot when beachBit 4 0 0](https://user-images.githubusercontent.com/39025953/124062368-332aea00-da63-11eb-9e9a-c289b97599d9.png)

**Comment by az0** [comment]
Yes, there was an improvement in BleachBit version 4.4.0 to significantly speed it up the first time it loads, and this changed the contents of the folder. I was worried the fonts might be missing, but they are there.

Please excuse my ignorance about how Simplified Chinese works, but my next concern is whether these fonts maybe lack the ability to show Simplified Chinese characters. I tried testing this by pasting the text 网页浏览器 (from BleachBit's Simplified Chinese translation) into Wordpad. Then I tried to change the font to Segoe UI or Tahoma, but the system did not allow a font change. The font always stayed as SimSun.

This is what I found so far, but maybe you have some insight.

**Comment by huastone2018** [comment]
Segoeui and tahoma can not support simplified Chinese display, them only support Western characters， such as english and Greek alphabet。 I hope the BleachBit developers can fix the probelm soon after. Thanks you very much @az0 too.

**Comment by abitrolly** [comment]
Yep, the proper fix would be to lookup system fonts if the necessary chars are missing for the current font. Good starting point to remind why Gtk is so slow in doing this.. because I already forgot it.

@huastone2018 which of default fonts on your system you can recommend to copy to BleachBit for display?

**Comment by rados** [comment]
@az0 
I can take this one.

**Comment by huastone2018** [comment]
@abitrolly the "msyh.ttf" is a good choice for me and it is a default font. “微软雅黑” is general used in Chinese, but it is not a fully free font and can be only freely used in MS windows. When I copy “微软雅黑” to fonts folder in BleachBit v4.4.0, display is normal and two files exist, i.e. msyh.ttf and msyhbd.ttf. It looks some better when only msyh.ttf left. 
If thinking over full  freedom, Siyuan fonts(such as SourceHanSerifSC-Medium.otf, not a default font) will be a better choice.
SourceHanSerifSC-Medium.otf can be downloaded from https://github.com/adobe-fonts/source-han-serif/raw/release/OTF/SourceHanSerifSC_EL-M.zip.

**Comment by rados** [comment]
> Yep, the proper fix would be to lookup system fonts if the necessary chars are missing for the current font.

Or we can add in the installer script just several additional fonts that would cover all of the cases? Like in this article:

https://dlsdc.com/blog/what-font-to-use-for-what-language/

**Comment by az0** [comment]
@huastone2018 

I made fixes that may help with Traditional Chinese and Simplified Chinese. (It also has a similar fix for Japanese.) Are you please willing to test? 

## How to test installer

1. Delete the BleachBit fonts folder `c:\program files(x86\bleachbit\share\fonts`
1. Download the installer [build 4.4.1.1973](https://ci.bleachbit.org/?prefix=dl/4.4.1.1973-master/) or newer from the [CI build site](https://ci.bleachbit.org/)
1. When starting the installer, choose Simplified or Traditional Chinese.
1. Finish installation
1. Start the application

During installation, it will detect that you selected Chinese, so it will copy `msyh.ttc` and `msyhbd.ttc`.

For a better test, please repeat, but on step 3, choose the other option. (If you chose Traditional the first time, please choose Simplified the second time.)

## How to test portable mode

Portable mode is a separate test. Also get build 4.4.1.1973, but choose the [portable zip file](https://ci.bleachbit.org/dl/4.4.1.1973-master/BleachBit-4.4.1.1973-portable.zip). Use it normally by unzipping all the files and running `bleachbit.exe`. During its first start, it should copy the needed fonts to its `share\fonts` folder (relatively to where the portable application is).

**Comment by huastone2018** [comment]
When I tested this portable mode version 4.4.1.1973 by zip files, according to the aforementioned instructions, garbled characters appeared on the bleabit interface, as shown in the picture below:
![blbt](https://user-images.githubusercontent.com/39025953/126645213-fe5e527a-b698-4c50-9cad-8cad6a88844b.png)

**Comment by huastone2018** [comment]
@az0
I cannot test the installed version of bleachbit because I cannot download it. When I clicked the aforementioned download link, the window that appeared did not show the download button.

**Comment by huastone2018** [comment]
The window screenshot:
![blbt](https://user-images.githubusercontent.com/39025953/126646445-e733525e-9581-48cc-81eb-901c239b61c2.png)

**Comment by Nevuly** [comment]
I got same problem in Windows 11 (64bit), and I using Korean. Texts are unreadable like him.

![bug02](https://user-images.githubusercontent.com/32860694/126855196-e6cd2d93-4fd2-4a50-8e22-585a90e0df38.png)

And I think it is not bug, but that string is too long, so It need to be more short..

![bug01](https://user-images.githubusercontent.com/32860694/126855207-55f6e3b8-2d00-4ea3-9dc0-e0efabaf62f0.png)

**Comment by az0** [comment]
@huastone2018 
You may not see the links if there is a JavaScript blocker because the content is generated dynamically. You could try another browser, or disable the JavaScript blocker, or here are the direct links

- [BleachBit-4.4.1.1975-portable.zip](https://ci.bleachbit.org/dl/4.4.1.1975-master/BleachBit-4.4.1.1975-portable.zip)
- [BleachBit-4.4.1.1975-setup.exe](https://ci.bleachbit.org/dl/4.4.1.1975-master/BleachBit-4.4.1.1975-setup.exe)

Please note that build 1975 has an additional fix for the Windows installer (from Rados) that is not in earlier builds.

@Nevuly 

Please test build 1975 (links above) or later. The newest builds are at the [BleachBit CI build site](https://ci.bleachbit.org/).

Ideally you would please test both the installer and portable version, as each has a different fix for the Japanese fonts. After testing, please add a comment here to let me know how it went.

**Comment by Nevuly** [comment]
@az0 I tested both version in test build, it still broken text.

I installed 4.4.1.1975.

1. Portable Version
![portable](https://user-images.githubusercontent.com/32860694/126856800-046f3950-0558-4968-950d-88911dcf08a3.png)

2. Installed Version
![normal](https://user-images.githubusercontent.com/32860694/126856805-73ea504e-d4d6-4c46-8117-d78c9c0ca46c.png)

P.S. I think you need to change fonts to malgun gothic (Windows standard font). It is work well in 4.4.0

**Comment by Nevuly** [comment]
In 4.4.0, I changed font file to malgun gothic, It is work. I think Japanese and Chinese need to use it too.

(I can't test it that two languages because bleachbit not support change languages)

![fixed1`](https://user-images.githubusercontent.com/32860694/126873970-a341d329-891c-4310-9ab8-267337139970.png)
![fixed2](https://user-images.githubusercontent.com/32860694/126873972-5bf203ef-97b2-4f0a-a3d1-78e72b832364.png)

**Comment by huastone2018** [comment]
@az0
I tested the installed version and it seems to be displayed normally, but the green version that is directly decompressed still does not work. Version is 4.4.1.1975.
Install version:
![b2](https://user-images.githubusercontent.com/39025953/127595735-490b076b-c05e-42bd-a9e9-133a9f653fe3.png)

Unzip version:
![b1](https://user-images.githubusercontent.com/39025953/127595762-92fd40d7-02ca-492b-b22e-317dcbfe68f8.png)

**Comment by az0** [comment]
@huastone2018 
Thank you for the excellent information. This will help troubleshoot and fix the issue. When another fix is ready, I will get back with you

**Comment by BraINstinct0** [comment]
Currently the broken(for Korean) version is on the website. Will try the newer installed version
Edit: still the same issue exist in 4.4.1976, installed version.
![image](https://user-images.githubusercontent.com/29201475/127724570-6ad27460-8b4b-4436-b5c5-7438f87a0eb4.png)

**Comment by Cazaea** [comment]
现在唯一能做的就是，从 "C:\Windows\Fonts" 搜索"Yahei"，将他们复制到 "C:\Program Files(x86)\BleachBit\share\fonts" , 然后选中字体，右键选择"Shred with BleachBit" ，问题可以解决。

The only thing you can do now is to search "Yahei" from "C: \ Windows \ Fonts", copy them to "C: \ Program Files (x86) \ Bleachbit \ Share \ Fonts", then select the font, right click "Shred with BleachBit", the problem can be solved.

**Comment by az0** [comment]
Today I made a few improvements:
- Add support for Korean
- Fix font copying in portable mode

These changes in version 4.4.1.1978 (or higher), which is available from the [BleachBit CI site](https://ci.bleachbit.org/). In case the site does not work normally, here are direct link

* [1978 portable (zip)](https://ci.bleachbit.org/dl/4.4.1.1978-master/BleachBit-4.4.1.1978-portable.zip)
* [1978 installer (exe)](https://ci.bleachbit.org/dl/4.4.1.1978-master/BleachBit-4.4.1.1978-setup.exe)

Next I will check whether it is feasible to link fonts instead of copying them.

**Comment by az0** [comment]
These are my notes about test on Windows 10 20H2 to use links to avoid copying fonts.

In general, non-admin user can create hardlinks, but the hardlinks cannot point to files under the c:\windows folder.  System will give "access is denied" error: this seems to be an undocumented security restriction.

Admin user can create symlinks to fonts in c:\windows\fonts, BleachBit (fontconfig/GTK) can use them for the GUI, and non-admin can delete the symlinks. Symlinks are more robust in case the folders are on different shares, which would happen with portable software on removable drive.

NSIS has no native functionality for creating either kind of link, while Python can do it since Python version 3.2.

**Comment by az0** [comment]
[Build 1982](https://ci.bleachbit.org/?prefix=dl/4.4.1.1982-master/) has more improvements 

* [1982 portable (zip)](https://ci.bleachbit.org/dl/4.4.1.1982-master/BleachBit-4.4.1.1982-portable.zip)
* [1982 installer (exe)](https://ci.bleachbit.org/dl/4.4.1.1982-master/BleachBit-4.4.1.1982-setup.exe)

**Comment by abitrolly** [comment]
@az0 can all the font collection logic be moved into separate module that can be imported and tested? So not just https://github.com/bleachbit/bleachbit/commit/19a4780676ed6870fa221fcf70124f4d855cae48 but including functions that test that fonts missing, explaining why fonts copying is needed (and linking to `fontconfig` issue with benchmarks that can be repeated and submitted upstream). Also, in NSIS do we copy the same fonts or just rebuild the cache for all that are available? How many fonts users can now choose that we copy a limited selection?

**Comment by abitrolly** [comment]
And especially interesting if we can copy symbols on demand with `fontconfig`. That means we can detect if a symbol is missing and do the lookup logic to find appropriate font.

**Comment by Nevuly** [comment]
I tested build 1982 both versions, Now korean show correctly without any broken.

![portable](https://user-images.githubusercontent.com/32860694/128678562-78b13efb-25b8-4f44-ae08-1d6e1ac95087.png)
![installer](https://user-images.githubusercontent.com/32860694/128678567-a23ce2d7-2e63-4771-bfe8-abab467fcb32.png)

**Comment by BraINstinct0** [comment]
Can confirm the above (@Nevuly).
Works like a charm on Windows 11 (Build 22000.120)

**Comment by rados** [comment]
@az0 
I also think that it is a good idea to move the functionality from NSIS to Python for the installed version as well. Even now with the existing functionality we will have several benefits:

- use of symlinks
- load of the system's language, not the installer's language as it is not mandatory to be the same
- load the system language when the user changes it after installation. With the current version we fail to load the system language if it is changed after installation even if we support it initially. For example if the user install under English and then changes to Korean.

**Comment by az0** [comment]
@Nevuly  @BraINstinct0 
Thank you for checking that it works on Korean! 

>  test that fonts missing
@abitrolly  
Do you mean missing because the copy failed or because because copy was not attempted?

 > Also, in NSIS do we copy the same fonts or just rebuild the cache for all that are available?
 
NSIS copies a smaller set of fonts, and the logic is not in sync. This is a WET programming problem.

NSIS could also call fc-cache.exe to rebuild the cache, so startup seems faster, but it does not.
 
 > How many fonts users can now choose that we copy a limited selection?
 
Could you rephrase the question? So far, NSIS supports three languages, and the portable version supports four. Sadly, there are many more that are not yet supported (e.g., Hebrew, Indonesian, Arabic). Rados posted a list that might help, but it could take more research to cross reference which font files are standard and preferred. For example, for Latin languages Times New Roman has been standard since maybe Windows 95 or earlier, but many modern applications now use Calibri. I am not an expert here, but that seems to be the case for Japanese with Meiyro and MS Gothic. I copied both.
 
 > can all the font collection logic be moved into separate module that can be imported and tested? 
 
 It could be tested the way it is coded in `Windows.py`, right?
 
 While working on this, I was thinking about moving the gettext code from `__init__.py` to a new `I18n.py` (internationalization) to make it easier to support dynamic language switching, which would make it easier to test the font problem. The font code could live there too.
  
 > And especially interesting if we can copy symbols on demand with fontconfig. That means we can detect if a symbol is missing and do the lookup logic to find appropriate font.
 
 I am not sure how to interact with fontconfig or check that a symbol is missing, but that would be helpful, if nothing else but to alert the user to a problem.
 
 > I also think that it is a good idea to move the functionality from NSIS to Python for the installed version as well.

@rados 
So you mean these two changes ?

1. Call the copy fonts function in Python on startup always on Windows, whether portable or installed
2. Adjust the GTK fonts target folder to be under the `%ProgramFiles` when the software is installed. (Specifically, it could be relative to `bleachbit.bleachbit_exe_path`.)

**Comment by rados** [comment]
@az0 
1. Yes.
2. Seems like it is not mandatory to have this folder on a specific place, we can access it as a relative to bleachbit_exe_path.

**Comment by abitrolly** [comment]
@az0 

> NSIS copies a smaller set of fonts, and the logic is not in sync. This is a WET programming problem.

Run a separate Python script from NSIS instead of coding everything in two places?

> NSIS could also call fc-cache.exe to rebuild the cache, so startup seems faster, but it does not.

And even if it will run, this won't help portable version, right?

>> How many fonts users can now choose that we copy a limited selection?
> Could you rephrase the question? 

I ask if BleachBit shows only copied languages when people try to change fonts?

> It could be tested the way it is coded in Windows.py, right?

Yes, but maintaining Windows font mapping is not a core value of BleachBit, so it could be better managed as a separate repo that people can easier reuse and contribute to speed up their Gtk+Python apps.

> I am not sure how to interact with fontconfig or check that a symbol is missing, but that would be helpful, if nothing else but to alert the user to a problem.

Me neither knows even what is the best place to ask. https://gitlab.gnome.org/GNOME/gtk/-/issues ?

**Comment by rados** [comment]
I think there is one more option. We can show information message / progress bar created with TKinter and close it after loading GTK.

**Comment by az0** [comment]
> I think there is one more option. We can show information message / progress bar created with TKinter and close it after loading GTK.

I like the creativity. However, this would require installing the TkInter libraries. On Ubuntu, the archive download size is 692 kB and installation size is 2,968 kB.

A lighter weight option might be the native Windows GUI via pywin32.

**Comment by az0** [comment]
@abitrolly 

> > NSIS copies a smaller set of fonts, and the logic is not in sync. This is a WET programming problem.
> 
> Run a separate Python script from NSIS instead of coding everything in two places?

Yes, that could work. It could be like `bleachbit_console.exe --copy-fonts`

> > NSIS could also call fc-cache.exe to rebuild the cache, so startup seems faster, but it does not.
> 
> And even if it will run, this won't help portable version, right?

Yes, it won't help portable, but it's a more intuitive experience because NSIS is already showing a progress bar, while BleachBit would not show any GUI yet while running fc-cache.exe.

> > > How many fonts users can now choose that we copy a limited selection?
> > > Could you rephrase the question?
> 
> I ask if BleachBit shows only copied languages when people try to change fonts?

What do you mean copied languages? 

When a user changes the system language (like English to Japanese) and restarts BleachBit, it should copy any new fonts that are missing.


> > It could be tested the way it is coded in Windows.py, right?
> 
> Yes, but maintaining Windows font mapping is not a core value of BleachBit, so it could be better managed as a separate repo that people can easier reuse and contribute to speed up their Gtk+Python apps.

😔 It is sad this is a common problem

**Comment by rados** [comment]
I have the following idea. For the upcoming quick version with fixes we can revert all fontconfig changes and add a native Windows control (message box or even a progress bar) that would inform users that the delay on the first run is okay and is under control. This way we won't cut languages and will have more time to make a comprehensive fix. What do you think?

**Comment by az0** [comment]
> I have the following idea. For the upcoming quick version with fixes we can revert all fontconfig changes and add a native Windows control (message box or even a progress bar) that would inform users that the delay on the first run is okay and is under control. This way we won't cut languages and will have more time to make a comprehensive fix. What do you think?

This sounds good if it can be done relatively quickly and doesn't add much size to the Windows installer. A progress bar is better than a message box.

**Comment by rados** [comment]
> > I have the following idea. For the upcoming quick version with fixes we can revert all fontconfig changes and add a native Windows control (message box or even a progress bar) that would inform users that the delay on the first run is okay and is under control. This way we won't cut languages and will have more time to make a comprehensive fix. What do you think?
> 
> This sounds good if it can be done relatively quickly and doesn't add much size to the Windows installer. A progress bar is better than a message box.

I've just implemented the splash screen part (PR #1263 not to be merged but just to show the idea) and just started to revert all fontconfig related changes. I realized that we can actually keep them so the fixed languages are still copied quickly. Also on the first run we can show a message box that asks the user if the font is OK and if it is not we can automatically run the full fontconfig cache after application restart. Or this is too complex?

**Comment by az0** [comment]
> I've just implemented the splash screen part (PR #1263 not to be merged but just to show the idea, it can be tested if the fontconfig cache is removed and the first_start property is set to True) and just started to revert all fontconfig related changes.

Thank you, Rados. I will take a look.

How is the font cache triggered to run in the background while the user watches the splash screen? 

> I realized that we can actually keep them so the fixed languages are still copied quickly. Also on the first run we can show a message box that asks the user if the font is OK and if it is not we can automatically run the full fontconfig cache after application restart. Or this is too complex?

This is interesting, but let's save this for a future release please. There could be some things to work out. For example, modifying the font configuration file (under c:\program files...) may require admin privileges, and in some cases (corporate laptop), user may not have admin privileges.

**Comment by rados** [comment]
> How is the font cache triggered to run in the background while the user watches the splash screen?

Splash is shown in a thread which ends when the Gtk starts loading the GUI after finishes the cache. The code is a bit hacky, I just wanted a quick running illustration.
